File: jssrc/StateModel.js
/**
 *  $Id: StateModel.js 2416 2007-11-08 17:35:41Z nick $
 */
function StateModel(initState, endState) {
	this.initState = initState;
	this.currentState = initState;
	this.finalState = endState;
	this.states = Array();
}

StateModel.prototype = {

	addState : function(state) {
		for(var i = 0; i <this.states.length; i++) {
			if(state.stateName == this.states[i].stateName) {
				throw 'Programming error: state "'+state.stateName+'" already exists in this model';
			}
		}
		this.states.push(state);
	},

	getStates : function() {
		return this.states;
	},

	getState : function() {
		return this.currentState;
	},

	getFinalState : function() {
		return this.finalState;
	},

	getCurrentState : function() {
		return this.currentState;
	},

	setState : function(state) {
		this.currentState = state;
	},
	
	/**
	 * This outputs in GraphViz 'dot' notation suitable for OmniGraffle and GraphViz
	 */
	toString : function() {
		var diagraph = 'digraph state_model{\narrowhead=normal;\nnode[shape=polygon,sides=8,color=red,regular=true];\n' + this.initState ;
		for(var i = 0; i < this.states.length; i++) {
			diagraph += this.states[i].toString();
		}
		diagraph += this.finalState;
		diagraph += '}\n';
		return diagraph;
	}
};

File: src/com/cortexity/fsm/StateModel.java
package com.cortexity.fsm;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

public class StateModel {

    private final Set<State> states = new HashSet<State>();

    private final State finalState;

    private State currentState;

    public StateModel(State initState, State endState) {
        currentState = initState;
        finalState = endState;
    }

    public void addState(State state) {
        this.states.add(state);
    }

    Collection<State> getStates() {
        return states;
    }

    public State getState() {
        return currentState;
    }

    State getFinalState() {
        return this.finalState;
    }

    State getCurrentState() {
        return this.currentState;
    }

    void setState(State state) {
        this.currentState = state;
    }

}
